nova/src/circuits/supernova/augmented.rs

```diff
@@ -0,0 +1,566 @@
+
+    _random_oracle: PhantomData<RO>,
+}
+
+impl<G1, G2, C1, C2, RO>
+    AllocVar<SuperNovaAugmentedCircuitInput<G1, G2, C1, C2, RO>, G1::ScalarField>
+    for SuperNovaAugmentedCircuitInputVar<G1, G2, C1, C2, RO>
+where
+    G1: SWCurveConfig,
+    G2: SWCurveConfig<BaseField = G1::ScalarField, ScalarField = G1::BaseField>,
+    G1::BaseField: PrimeField,
+    G2::BaseField: PrimeField,
+    C1: CommitmentScheme<Projective<G1>>,
+    C2: CommitmentScheme<Projective<G2>>,
+    RO: SpongeWithGadget<G1::ScalarField>,
+    RO::Var: CryptographicSpongeVar<G1::ScalarField, RO, Parameters = RO::Config>,
+{
+    fn new_variable<T: Borrow<SuperNovaAugmentedCircuitInput<G1, G2, C1, C2, RO>>>(
+        cs: impl Into<Namespace<G1::ScalarField>>,
+        f: impl FnOnce() -> Result<T, SynthesisError>,
+        mode: AllocationMode,
+    ) -> Result<Self, SynthesisError> {
+        let ns = cs.into();
+        let cs = ns.cs();
+
+        let input = f()?;
+        let input = input.borrow();
+
+        let input = match input {
+            SuperNovaAugmentedCircuitInput::Base { vk, z_0, pc, num_augmented_circuits } => {
+                let U = vec![None; *num_augmented_circuits];
+                let U_secondary = vec![None; *num_augmented_circuits];
+                let u = R1CSInstance::<G1, C1> {
+                    commitment_W: C1::Commitment::default(),
+                    X: vec![G1::ScalarField::ONE; AUGMENTED_CIRCUIT_NUM_IO],
+                };
+                SuperNovaAugmentedCircuitNonBaseInput {
+                    vk: *vk,
+                    i: G1::ScalarField::ZERO,
+                    pc_prev: *pc,
+                    pc: *pc,
+                    z_0: z_0.clone(),
+                    z_i: z_0.clone(),
+                    U,
+                    U_secondary,
+                    u,
+                    proof: NIMFSProof::default(),
+                }
+            }
+            SuperNovaAugmentedCircuitInput::NonBase(non_base) => non_base.clone(),
+        };
+
+        let vk = FpVar::new_variable(cs.clone(), || Ok(input.vk), mode)?;
+        let i = FpVar::new_variable(cs.clone(), || Ok(input.i), mode)?;
+        let pc_prev = FpVar::new_variable(
+            cs.clone(),
+            || Ok(G1::ScalarField::from(input.pc_prev)),
+            mode,
+        )?;
+        let pc = FpVar::new_variable(cs.clone(), || Ok(G1::ScalarField::from(input.pc)), mode)?;
+        let z_0 = input
+            .z_0
+            .iter()
+            .map(|z| FpVar::new_variable(cs.clone(), || Ok(z), mode))
+            .collect::<Result<_, _>>()?;
+        let z_i = input
+            .z_i
+            .iter()
+            .map(|z| FpVar::new_variable(cs.clone(), || Ok(z), mode))
+            .collect::<Result<_, _>>()?;
+
+        let U = input
+            .U
+            .iter()
+            .map(|U| {
+                let U = U.clone().unwrap_or_else(|| RelaxedR1CSInstance {
```

Comment:
why is this clone necessary? in the nova ivc impl, `U` is moved here.

Reply:
Because it's either `&input.U[i]` or default instance. Clone can be avoided with `std::borrow::Cow`, but I chose simplicity.